### PR TITLE
Use separate executor for each coordinator duty group

### PR DESCRIPTION
**Pull request by kfaraz**
### Motivation

- Some coordinator duties such as `CompactSegments` or `KillUnusedSegments` can be slow due to slowness on the Overlord or the metadata store.
- This causes the coordinator executor to get blocked up thus delaying segment loading and hand-offs, which in turn causes ingestion failures.
- With the introduction of custom coordinator duties, it becomes imperative to separate the executors to ensure that the core coordinator logic is never blocked.

### Changes
- Use separate executor for every duty group
- This change is thread-safe as every duty group uses its own copy of `DruidCoordinatorRuntimeParams` and does not share any other mutable instances with other duty groups.
- With the exception of `HistoricalManagementDuties`, duty groups are typically not very compute intensive and mostly perform database or HTTP I/O. So, coordinator resources would still mostly be available for `HistoricalManagementDuties`.

### Alternative approach
Have 1 executor for `HistoricalManagementDuties` and 1 for all other duty groups.
However, this still allows users to schedule some custom duties that compete with say the `IndexingServiceDuties` or `MetadataStoreManagementDuties`.

### Testing

Tested locally on a small cluster with `coordinator.period`, `indexingPeriod` and `metadataStoreManagementPeriod`, all set to `PT5S`. Compaction, coordination working as expected.

---

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [x] been tested in a test Druid cluster.

### CODE REVIEW

**Review by findingrish — APPROVED** [review]
LGTM!

File: server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java [hunk 1]
```diff
@@ -395,7 +396,7 @@ public void stop()
 
       started = false;
 
-      exec.shutdownNow();
+      stopAllDutyGroupExecutors();
```
[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java:399]: No, duties do not run and metrics are not emitted if the coordinator is not leader.

See https://github.com/apache/druid/blob/3a75ee2098390922501e0993b8c29e9a85c42b3b/server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java#L508-L515

But I guess it is better to call `stopAllDutyGroupExecutors` in `stopBeingLeader` itself.
I don't want to run the whole CI just for this as it doesn't really make a difference. I will address it in a follow up PR, unless there are other comments in the current PR.

[the remark this answers is not in the document]
Comment (AmatyaAvadhanula) [reply server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java:399]: Sorry, I meant to ask about (long running) duties that had already been scheduled and are waiting to be run.
+1 from me as I have no other comments

**Review by AmatyaAvadhanula — APPROVED** [review]

File: server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kfaraz) [reply server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java:399]: Yes, a duty group which has been scheduled for later will not run either.

https://github.com/apache/druid/blob/3a75ee2098390922501e0993b8c29e9a85c42b3b/server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java#L701-L704

A duty group which has already started will not run any more duties after the currently running one has finished.
https://github.com/apache/druid/blob/3a75ee2098390922501e0993b8c29e9a85c42b3b/server/src/main/java/org/apache/druid/server/coordinator/DruidCoordinator.java#L747-L749

**Comment by kfaraz** [comment]
Thanks for the reviews, @findingrish , @AmatyaAvadhanula !

**Pull request merged**